homeassistant/components/oralb/__init__.py

```diff
@@ -1,30 +1,104 @@
 
 # For your initial PR, limit it to 1 platform.
 PLATFORMS: list[Platform] = [Platform.SENSOR]
+_LOGGER = logging.getLogger(__name__)
 
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up OralB from a config entry."""
+    address: str = entry.data[CONF_MAC]
+    print("using address: ", address)
+    ble_device = bluetooth.async_ble_device_from_address(hass, address.upper(), True)
+    if not ble_device:
+        raise ConfigEntryNotReady(f"Could not find OralB device with address {address}")
+
+    orlb = OralB(ble_device)
+
+    @callback
+    def _async_update_ble(
+        service_info: bluetooth.BluetoothServiceInfoBleak,
+        change: bluetooth.BluetoothChange,
+    ) -> None:
+        """Update from a ble callback."""
+        orlb.set_ble_device(service_info.device)
+
+    entry.async_on_unload(
+        bluetooth.async_register_callback(
+            hass,
+            _async_update_ble,
+            BluetoothCallbackMatcher({ADDRESS: address}),
+            bluetooth.BluetoothScanningMode.PASSIVE,
+        )
+    )
+
+    async def _async_update():
+        """Update the device state."""
+        await orlb.gatherdata()
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name=DOMAIN,
+        update_method=_async_update,
+        update_interval=timedelta(seconds=UPDATE_SECONDS),
+    )
 
-    hass.data.setdefault(DOMAIN, {})
-    # hass.data[DOMAIN][entry.entry_id] = MyApi(...)
+    try:
+        async with async_timeout.timeout(DEVICE_TIMEOUT):
+            await coordinator.async_config_entry_first_refresh()
+    except asyncio.TimeoutError as ex:
+        raise ConfigEntryNotReady(
+            "Unable to communicate with the device; "
+            f"Try moving the Bluetooth adapter closer to {DOMAIN}"
+        ) from ex
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = OralBData(
+        entry.title, orlb, coordinator
+    )
 
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    entry.async_on_unload(entry.add_update_listener(_async_update_listener))
+
+    async def _async_stop(event: Event) -> None:
+        """Close the connection."""
+        orlb.disconnect()
 
+    entry.async_on_unload(
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _async_stop)
+    )
     return True
 
 
+async def _async_update_listener(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Handle options update."""
+    print("update listener")
+    data = hass.data[DOMAIN][entry.entry_id]
+    data.device.gatherdata()
+    print(data.device.result)
+
+
 async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Unload a config entry."""
-    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
-        hass.data[DOMAIN].pop(entry.entry_id)
-
-    return unload_ok
+    data = hass.data[DOMAIN][entry.entry_id]
+    data.device.disconnect()
```

Comment:
We should never keep a connection open. There are only limited connections available, especially via the Bluetooth proxies. We shouldn't dedicate one to just 1 device.

Reply:
The connection is not kept alive because from the package it disconnects using the `data.device.disconnect()`.
Also the OralB toothbrushes don't stay on more than a few seconds after you finish brushing, so I don't think it's gonna be a problem with the limit of bluetooth devices.